### ISSUE TITLE
[Legacy codegen] Higher order bits not sign-extended in signed integer array in storage that may be accessed via inline assembly

### ISSUE THREAD

**Issue by bshastry** [ethereum/solidity#13180]
In the semantic test below, legacy codegen does not sign-extend the second element of the storage array but via-ir does. This may be accessed via inline assembly. Bug similar to https://github.com/ethereum/solidity/pull/12050

Note that Solidity access is just fine (i.e., sign-extension happens correctly via legacy or IR). Also note that the value returned by the call to `g()` in the test below is via legacy (i.e., the non-sign-extended value).

```
contract C0 {
  int192[2]   s3;
  constructor(int192[2] memory i0)   {
    s3 = i0;
  }
  function f() external returns (int192[2] memory o) { o = s3; }
  function g() external returns (uint o) { assembly { o := sload(add(s3.slot, 1)) } }
}
// ====
// compileViaYul: also
// ----
// constructor(): 0x0000000000000000000000000000000000000000000000000000000000000000, 0xFFFFFFFFFFFFFFFFF6F6BEA5083DFED985055AB0D43C9DB425C40FEAC74632E5 ->
// gas irOptimized: 136872
// gas legacy: 179142
// gas legacyOptimized: 135428
// f() -> 0, -221565838963264769260736510461167584867121330438731844891
// g() -> 6055535896423415994575052912746498831235234114025302668005
```

**Comment by ekpyron** [comment]
Simpler case:
```
contract C {
  int136[1] s;
  function test() external returns (uint o) {
	s = [int136(-1)];
	assembly { o := sload(s.slot) }
  }
}
// ====
// compileViaYul: also
// ----
// test() -> 0xffffffffffffffffffffffffffffffffff
// test() -> -1
```
(the first expectation is legacy, the second via-IR)

So the cases are storage arrays in which only a single value is stored per slot (i.e. ``int8`` would have consistent behaviour - and since that's packed, the values in storage cannot be sign-extended at all).

While we should probably document this difference, I think it's basically harmless, given that signextending cleanup is performed on solidity reads.

However, a second pair of eyes on this may be nice to see, if anyone can come up with a case in which this would cause actual issues.

**Comment by bshastry** [comment]
> So the cases are storage arrays in which only a single value is stored per slot

Ah I see, this would mean cases where bit width is `> 128`.

> I think it's basically harmless, given that signextending cleanup is performed on solidity reads.

Isn't this similar to https://github.com/ethereum/solidity/blob/7e7c68e5bf910b12ee9326dd34926c61261a4a71/test/libsolidity/semanticTests/immutable/immutable_signed.sol#L5-L9

Is non cleaned up value that is only accessible via inline assembly okay?

**Comment by ekpyron** [comment]
It's borderline and will probably require some thought on a case-by-case basis...
For storage, in a sense, the "default" is actually *not* to sign-extend smaller types, due to the packed layout for ``<= 128`` bytes types...
I'd myself also not necessarily have considered the immutable case an "important bug", so that was also at least borderline.

**Comment by ekpyron** [comment]
But yeah, I'd be happy about more opinions on this.

**Comment by ekpyron** [comment]
I'm staging this for discussion next Wednesday (even though I'm off there, but maybe that results in some more opinions anyways).

My current thinking on this would be to (1) document this properly (2) take care of optimizing the Yul optimizer :-) - and (3) drop legacy codegen ;-) - that'd solve the issue eventually :-).

**Comment by bshastry** [comment]
My two cents: If the test case does not really affect real-world contracts, it is safe to just document this properly and eventually not have to care about it when we deprecate legacy codegen. OTOH, if there is no way of finding out real-world impact and there is a small possibility that there may be an impact, I would tend to just be consistent with the other linked bug (signed immutables) and mark it a low-priority bug.

**Comment by NunoFilipeSantos** [comment]
@bshastry what is the impact of this bug (High, Medium, Low)?

**Comment by bshastry** [comment]
I would say low because I have not been able to find smart contracts that read from a storage array via inline assembly.

**Comment by chriseth** [comment]
Isn't this documented in https://docs.soliditylang.org/en/v0.8.17/ir-breaking-changes.html#cleanup ? We could clarify that this also applies to storage.

**Comment by NunoFilipeSantos** [comment]
@bshastry is @chriseth correct here?

**Comment by bshastry** [comment]
Yes, @chriseth is right. Documenting the divergence for storage arrays should be sufficient.

**Comment by ekpyron** [comment]
Demoting this from a codegen error to a documentation issue. We should still definitely fix it, it's bad if this changes without proper documentation.
